File: Leetcode/1448.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, int> dp;   // dp[u] = the maxvalue in the path from u to root
    unordered_map<TreeNode*, TreeNode*> fa;
    int ans;
    void dfs(TreeNode* root, TreeNode* father_of_root) {
        if (root) {
            fa[root] = father_of_root;
            dfs(root->left, root);
            dfs(root->right, root);
        }
    }
    int solve(TreeNode* u) {
        if (dp.find(u) != dp.end()) {
            return dp[u];
        }
        if (fa[u]) {
            dp[u] = max(u->val, solve(fa[u]));
        }
        else {
            dp[u] = u->val;
        }
        return dp[u];
    }
    void dfs2(TreeNode* u) {
        if (u) {
            if (solve(u) <= u->val) {
                ++ans;
            }
            dfs2(u->left);
            dfs2(u->right);
        } 
    }
    int goodNodes(TreeNode* root) {
        dp.clear();
        fa.clear();
        dfs(root, nullptr);
        ans = 0;
        dfs2(root);
        return ans;
    }
};

File: Leetcode/1448.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var (
	dp map[*TreeNode]int
	fa map[*TreeNode]*TreeNode
	ans int
)

func dfs1(u *TreeNode, father *TreeNode) {
	if u != nil {
		fa[u] = father
		dfs1((*u).Left, u)
		dfs1((*u).Right, u)
	}
}

func solve(root *TreeNode) int {
	val, ok := dp[root]
	if ok {
		return val
	}
	if fa[root] == nil {
		dp[root] = (*root).Val
	} else {
		var temp = solve(fa[root])
		if temp > (*root).Val {
			dp[root] = temp
		} else {
			dp[root] = (*root).Val
		}
	}
	return dp[root]
}

func dfs2(root *TreeNode) {
	if root != nil {
		if (*root).Val >= solve(root) {
			ans++
		}
		dfs2((*root).Left)
		dfs2((*root).Right)
	}
}

func goodNodes(root *TreeNode) int {
	if root == nil {
		return 0
	}
	dp = nil
	fa = nil
	dp = make(map[*TreeNode]int)
	fa = make(map[*TreeNode]*TreeNode)
	dfs1(root, nil)
	ans = 0
	dfs2(root)
	return ans
}